python/ray/train/_internal/syncer.py

```diff
@@ -0,0 +1,583 @@
+import abc
+import threading
+import traceback
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Union,
+    Optional,
+    Tuple,
+)
+import warnings
+
+import logging
+import time
+from dataclasses import dataclass
+
+
+from ray._private.thirdparty.tabulate.tabulate import tabulate
+from ray.air._internal.remote_storage import (
+    fs_hint,
+    upload_to_uri,
+    download_from_uri,
+    delete_at_uri,
+    is_non_local_path_uri,
+)
+from ray.util import log_once
+from ray.util.annotations import PublicAPI, DeveloperAPI
+from ray.widgets import Template
+
+
+logger = logging.getLogger(__name__)
+
+# Syncing period for syncing checkpoints between nodes or to cloud.
+DEFAULT_SYNC_PERIOD = 300
+
+# Default sync timeout after which syncing processes are aborted
+DEFAULT_SYNC_TIMEOUT = 1800
+
+_DEPRECATED_VALUE = "DEPRECATED"
+
+
+@PublicAPI
+@dataclass
+class SyncConfig:
+    """Configuration object for Train/Tune file syncing o `RunConfig(storage_path)`.
+
+    See :ref:`tune-persisted-experiment-data` for an overview of what data is
+    synchronized.
+
+    The Tune driver (on the head node) syncs the experiment directory to the cloud
+    (which includes experiment state such as searcher state, the list of trials
+    and their statuses, and trial metadata).
+
+    See :ref:`tune-storage-options` for more details and examples.
+
+    Args:
+        sync_period: Minimum time in seconds to wait between two sync operations.
+            A smaller ``sync_period`` will have more up-to-date data at the sync
+            location but introduces more syncing overhead.
+            Defaults to 5 minutes.
+        sync_timeout: Maximum time in seconds to wait for a sync process
+            to finish running. This is used to catch hanging sync operations
+            so that experiment execution can continue and the syncs can be retried.
+            Defaults to 30 minutes.
+    """
+
+    upload_dir: Optional[str] = _DEPRECATED_VALUE
+    syncer: Optional[Union[str, "Syncer"]] = _DEPRECATED_VALUE
+    sync_period: int = DEFAULT_SYNC_PERIOD
+    sync_timeout: int = DEFAULT_SYNC_TIMEOUT
+    sync_artifacts: bool = _DEPRECATED_VALUE
```

Comment:
Should we actually keep this config?

Reply:
Decision: for 2.7, the semantics of this should be "whether remote workers (or the trial driver, if the trial is non-distributed) should sync artifacts to remote storage on each report call, subject to sync_period?".

False by default in 2.7.

Reply:
I think we should also keep `sync_on_checkpoint`, which seems like an important capability.

Reply:
I think `sync_artifacts_on_checkpoint` makes more sense as a param name -- even though it's different. Lmk what you think.

I think the old `sync_on_checkpoint` semantics were so different that it's worth making a switch here.

Reply:
`sync_artifacts_on_checkpoint` is unused at the moment. Let's land this PR first, then merge it into https://github.com/ray-project/ray/pull/38871 and switch `sync_on_checkpoint` to `sync_artifacts_on_checkpoint`.

Reply:
Actually, let's just do that one first. I'll update this tomorrow.